core/shared/src/main/scala/org/http4s/Service.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2013 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+
+import cats.Alternative
+import cats.Applicative
+import cats.Monad
+import cats.StackSafeMonad
+import cats.effect.kernel.Resource
+import cats.mtl.Local
+
+import Service._
+
+sealed abstract class Service[F[_], A, B] { self =>
+  def applyOrElse(a: A, default: A => Resource[F, B]): Resource[F, B]
+
+  def map[C](f: B => C): Service[F, A, C] =
+    new Service[F, A, C] {
+      def applyOrElse(a: A, default: A => Resource[F, C]): Resource[F, C] = {
+        val rb = self.applyOrElse(a, checkFallback)
+        if (isFallback(rb)) default(a) else rb.map(f)
+      }
+    }
```

Comment:
Maybe I can lambda-lift these because they're SAMs and make them less noisy? :shrug:

Reply:
Using SAM is nice, at lease to remove the `new Service[F, A, C]` constructor line.

Reply:
I tried and couldn't make it compile.  Does it work with type parameters?

Reply:
Hm, it should work 🤔 let me take a look.

Reply:
Seems like SAMs don't work for `sealed`.

Reply:
Perhaps we can drop `sealed` in favour of a `private[http4s]` constructor?

Reply:
The encoding doesn't work yet either.  That PartialFunction trick may not be viable here.

Reply:
I'm trying that right now, but it also doesn't seem to work 🤔 

Would it be so bad if it became a public trait?

Reply:
If what did?  The Service?

I suspect we can optimize it with an algebra and an interpreter that "compiles" it to a function.  But whatever makes it actually work now.

Reply:
Right, let's not worry about SAMs until we decide the final encoding :)

Reply:
I was curious whether a SAM would affect performance.  Doesn't seem to.

Reply:
Right, SAMs should not affect performance. Lambdas are ultimately implemented as classes; for SAMs this class is synthesized at runtime and thus bring a small saving in bytecode.